packages/utils/src/schema/sanitizeDataForNewSchema.ts

```diff
@@ -0,0 +1,208 @@
+  StrictRJSFSchema,
+  ValidatorType,
+} from "../types";
+import { PROPERTIES_KEY, REF_KEY } from "../constants";
+import retrieveSchema from "./retrieveSchema";
+
+/** Sanitize the `data` associated with the `oldSchema` so it is considered appropriate for the `newSchema`. If the new
+ * schema does not contain any properties, then `undefined` is returned to clear all the form data. Due to the nature
+ * of schemas, this sanitization happens recursively for nested objects of data. Also, any properties in the old schema
+ * that are non-existent in the new schema are set to `undefined`. The data sanitization process has the following flow:
+ *
+ * - If the new schema is an object that contains a `properties` object then:
+ *   - Create a `removeOldSchemaData` object, setting each key in the `oldSchema.properties` having `data` to undefined
+ *   - Create an empty `nestedData` object for use in the key filtering below:
+ *   - Filter each key in the `newSchema.properties` storing the filtered keys in `keysToKeep` as follows:
+ *     - Get the `formValue` of the key from the `data`
+ *     - Get the `oldKeySchema` and `newKeyedSchema` for the key, defaulting to `{}` when it doesn't exist
+ *     - Retrieve the schema for any refs within each `oldKeySchema` and/or `newKeySchema`
+ *     - Get the types of the old and new keyed schemas and if the old doesn't exist or the old & new are the same then:
+ *       - If `removeOldSchemaData` has an entry for the key, delete it since the new schema has the same property
+ *       - If type of the key in the new schema is `object`:
+ *         - Store the value from the recursive `sanitizeDataForNewSchema` call in `nestedData[key]`
+ *       - Otherwise, return true if the new schema and its associated data passes all of these conditions:
+ *         - The schema has a falsey `readOnly` property value OR
+ *         - The schema has a falsey `default` property OR
+ *         - The `default` property value is the same as the `formValue`
+ *     - Unless true was returned above, return `false` to not return the current key in the filter
+ *   - Once all keys have been processed, return an object built as follows:
+ *     - `{ ...removeOldSchemaData, ...nestedData, ...pick(data, keysToKeep) }`
+ * - Otherwise return `undefined`
+ *
+ * @param validator - An implementation of the `ValidatorType` interface that will be used when necessary
+ * @param rootSchema - The root JSON schema of the entire form
+ * @param [newSchema] - The new schema for which the data is being sanitized
+ * @param [oldSchema] - The old schema from which the data originated
+ * @param [data={}] - The form data associated with the schema, defaulting to an empty object when undefined
+ * @returns - The new form data, with all the fields uniquely associated with the old schema set
+ *      to `undefined`. Will return `undefined` if the new schema is not an object containing properties.
+ */
+export default function sanitizeDataForNewSchema<
```

Comment:
I feel like this is different than omitExtraData in that when we change from one schema to another we are trying to remove anything that doesn't match the old schema AND keep anything that does. The old implementation just removed things, especially objects since the nesting of data wasn't checked.

Reply:
It is different as `omitExtraData` only happens on submit